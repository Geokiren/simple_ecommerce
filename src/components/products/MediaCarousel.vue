<template>
  <div class="media-slider flex flex-col gap-1 w-full">
    <div class="selected-media w-full">
      <div class="relative aspect-video bg-secondary rounded-lg">
        <OptimizedImage 
          v-if="activeImage"
          :key="activeImage"
          class="w-full h-full object-cover"  
          :src="activeImage" 
          alt="Product Image" 
          @load="onImageLoad" 
        />
      </div>
    </div>

    <div v-if="areImagesGreaterThanOne" class="thumbnails flex gap-1 w-full h-16 lg:h-24 4xl:h-32 overflow-auto">
      <OptimizedImage 
        v-for="(image, index) in images" 
        :key="image" 
        class="cursor-pointer rounded-lg  bg-secondary" 
        :class="{ 'border-2 border-active': image === activeImage }" 
        :src="image"
        :alt="`thumbnail-${index}`"
        loading="lazy"
        @click="activeImage = image" 
      /> 
    </div>
  </div>
</template>


<script setup lang="ts">
  import { ref, computed, defineProps, onMounted, onUpdated } from 'vue';

  import OptimizedImage from '../shared/OptimizedImage.vue';

  interface Props {
    images: string[];
  }

  const { images } = defineProps<Props>();

  const imageLoaded = ref<boolean>(false);

  const onImageLoad = () => {
    imageLoaded.value = true;
  }

  const activeImage = ref('');

  const areImagesGreaterThanOne = computed(() => {
    return images.length > 1;
  });

  onMounted(() => {
    activeImage.value = images[0];
  });

  onUpdated(() => {
    activeImage.value = images[0];
  });
</script>
