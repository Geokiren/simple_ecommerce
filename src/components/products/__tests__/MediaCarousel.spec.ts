import { describe, expect, it } from 'vitest';
import { mount } from '@vue/test-utils';
import { nextTick } from 'vue';

import MediaCarousel from '../MediaCarousel.vue';

describe('MediaCarousel', () => {
  it('renders the selected media image', async () => {
    const images = ['image1.jpg', 'image2.jpg', 'image3.jpg'];
    const wrapper = mount(MediaCarousel, {
      props: { images }
    });

    await nextTick();

    expect(wrapper.find('.selected-media img').exists()).toBe(true);
    expect(wrapper.find('.selected-media img').attributes('data-src')).toBe(images[0]);
  });

  it('renders thumbnails when there are multiple images', async () => {
    const images = ['image1.jpg', 'image2.jpg', 'image3.jpg'];
    const wrapper = mount(MediaCarousel, {
      props: { images }
    });

    expect(wrapper.find('.thumbnails').exists()).toBe(true);
    expect(wrapper.findAll('.thumbnails img').length).toBe(images.length);
  });

  it('does not render thumbnails when there is only one image', async () => {
    const images = ['image1.jpg'];
    const wrapper = mount(MediaCarousel, {
      props: { images }
    });

    expect(wrapper.find('.thumbnails').exists()).toBe(false);
  });

  it('matches snapshot', async () => {
    const images = ['image1.jpg', 'image2.jpg', 'image3.jpg'];
    const wrapper = mount(MediaCarousel, {
      props: { images }
    });

    expect(wrapper.element).toMatchSnapshot();
  });
});
