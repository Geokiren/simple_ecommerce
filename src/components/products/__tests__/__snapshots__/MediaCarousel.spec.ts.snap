// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MediaCarousel > matches snapshot 1`] = `
<div
  class="media-slider flex flex-col gap-1 w-full"
>
  <div
    class="selected-media w-full"
  >
    <div
      class="relative aspect-video bg-secondary rounded-lg"
    >
      <!--v-if-->
    </div>
  </div>
  <div
    class="thumbnails flex gap-1 w-full h-16 lg:h-24 4xl:h-32 overflow-auto"
  >
    
    <div
      class="relative overflow-hidden cursor-pointer rounded-lg bg-secondary"
      loading="lazy"
    >
      <img
        alt="thumbnail-0"
        class="inset-0 w-full h-full object-cover transition-opacity duration-300 opacity-0"
        data-src="image1.jpg"
        src=""
      />
      <!--v-if-->
      <!--v-if-->
    </div>
    <div
      class="relative overflow-hidden cursor-pointer rounded-lg bg-secondary"
      loading="lazy"
    >
      <img
        alt="thumbnail-1"
        class="inset-0 w-full h-full object-cover transition-opacity duration-300 opacity-0"
        data-src="image2.jpg"
        src=""
      />
      <!--v-if-->
      <!--v-if-->
    </div>
    <div
      class="relative overflow-hidden cursor-pointer rounded-lg bg-secondary"
      loading="lazy"
    >
      <img
        alt="thumbnail-2"
        class="inset-0 w-full h-full object-cover transition-opacity duration-300 opacity-0"
        data-src="image3.jpg"
        src=""
      />
      <!--v-if-->
      <!--v-if-->
    </div>
    
  </div>
</div>
`;
